src/types/string.rs

```diff
@@ -87,10 +88,29 @@ impl PyString {
     /// Unpaired surrogates invalid UTF-8 sequences are
     /// replaced with U+FFFD REPLACEMENT CHARACTER.
     pub fn to_string_lossy(&self) -> Cow<str> {
-        // TODO: Handle error of `as_bytes`
-        // see https://github.com/PyO3/pyo3/pull/634
-        let bytes = self.as_bytes().unwrap();
-        String::from_utf8_lossy(bytes)
+        match self.to_string() {
+            Ok(s) => s,
+            Err(_) => {
+                unsafe {
+                    let py_bytes = ffi::PyUnicode_AsEncodedString(
+                        self.0.as_ptr(),
+                        CStr::from_bytes_with_nul(b"utf-8\0").unwrap().as_ptr(),
+                        CStr::from_bytes_with_nul(b"surrogatepass\0")
+                            .unwrap()
+                            .as_ptr(),
+                    );
+                    // Since we have a valid PyString and replace any surrogates, assume success.
+                    debug_assert!(!py_bytes.is_null());
+                    // ensure DECREF will be called
+                    self.py().from_owned_ptr::<PyBytes>(py_bytes);
```

Comment:
How about `crate::gil::register_pointer`?(My snippet was incorrect, sorry)

Reply:
Ok, done. What's the difference?

Reply:
1. Less overhead.
2. Our GIL pool has 3 types of buffer(owned ref, borrowed ref, pointer). 
In this case, we can use both owned and pointer buffer.
But pointer buffer is preferable for the case `GILGuard` drops earlier(I'll add an example but have no time now :sweat: ).